### PR TITLE
[SPARK-12915]  [SQL] add SQL metrics of numOutputRows for whole stage codegen

### PR DESCRIPTION
**Pull request by davies**
This PR add SQL metrics (numOutputRows) for generated operators (same as non-generated), the cost is about 0.2 nano seconds per row.

gen metrics

### CODE REVIEW

**Comment by davies** [comment]
cc @zsxwing

**Comment by rxin** [comment]
Can you document in the pull request description what gets added? For example, do we show cardinality for aggregations or joins?

**Comment by davies** [comment]
@rxin Aggregation and joins have the same numOutputRows as without codegen, only datasize and spillsize are not included in this PR, because I think they are not implemented correctly (need more thinking on that).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 1]
```diff
@@ -43,6 +43,17 @@ trait CodegenSupport extends SparkPlan {
   }
 
   /**
+    * Create a term for LongSQLMetricValue with `name`.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]:
```
Creates a metric using the specified name.

@returns name of the variable representing the metric
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala [hunk 2]
```diff
@@ -316,6 +327,10 @@ case class WholeStageCodegen(plan: CodegenSupport, children: Seq[SparkPlan])
     }
   }
 
+  private[sql] override def resetMetrics(): Unit = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: move this closer to the other metric function?

**Comment by rxin** [comment]
Sorry I'm still not getting the picture. Can you just paste a screenshot for the metrics?

**Comment by rxin** [comment]
The part I'm confused about is that we seem to have multiple numOutputRows in a single WholeStageCodegen operator?

**Comment by davies** [comment]
@rxin There will be a numOutputRows for some of the operators inside WholeStageCodegen, as the picture showed, because we still want to know the runtime metrics of individual operators.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala (hunk 2 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala]: There are in different classes.

**Comment by rxin** [comment]
LGTM

**Comment by rxin** [comment]
I'm going to merge this.

**Pull request closed without merging**